Testing.Rmd

```diff
@@ -442,159 +769,1106 @@ It's hard to give good general advice about writing tests, but you might find th
+* Top-level code interspersed in the test files that, e.g., defines test
+  objects or attaches packages with `library()`
+  
+These practices can undermine the quality of both the test suite and the interactive workflow:
+
+* They change the R landscape: `library()` alters the search path and
+  top-level assignments add objects to the current environment. This means the
+  circumstances under which you are testing do not necessarily reflect the
+  circumstances under which your package will be used. This makes it easier to
+  create subtle test bugs, which you will have to unravel in the future.
+* There's no official gesture, in devtools, testthat, or the RStudio IDE, for
+  ferreting out all these bits of code and running them when you want to work
+  on your tests.
+  Therefore you have to find and run this code "by hand".
+  This sort of *ad hoc* code creates hard-to-see dependencies between
+  different pieces of your testing suite and adds mystery when you least need
+  it: when you're trying to figure out why tests are failing.
+
+What should you do instead?
+When working on your tests, use `devtools::load_all()`, just like you do when working below `R/`.
+By default, `load_all()` does all of these things:
+
+* Simulates re-building, re-installing, and re-loading your package.
+* Makes everything in your package's namespace available, including unexported
+  functions and objects and anything you've imported from another package.
+* Attaches testthat, i.e. does `library(testthat)`.
+* Runs test helper files, i.e. executes `test/testthat/helper.R`.
+
+This eliminates the need for any `library()` calls below `tests/testthat/`, for the vast majority of R packages.
+Any instance of `library(testthat)` is clearly no longer necessary.
+Likewise, any instance of attaching one of your dependencies via `library(somePkg)` is unnecessary.
+In your tests, if you need to call functions from somePkg, do it just as you do below `R/`.
+If you have imported the function into your namespace, use `fun()`.
+If you have not, use `somePkg::fun()`.
+It's fair to say that `library(somePkg)` in the tests should be about as rare as taking a dependency via `Depends`, i.e. there is usually a better alternative.
+
+Test toubleshooting nirvana looks like this:
+In a fresh R sesion, you can do `devtools::load_all()` and immediately run an individual test or walk through it line-by-line.
+There is no need to hunt around for code that has to be run first, that is found elsewhere in the test file or in a different file altogether (i.e. some file that is not executed by `load_all()`).
+
+One other function that basically should never appear below `tests/testhat/` is `source()`.
+There are several special files with an official role in testthat workflows (see below), not to mention the entire R package machinery, that provide better ways to make functions, objects, and other logic available in your tests.
+
+## Files relevant to testing  {#tests-files-overview}
```

Comment:
unsurprisingly I am thinking about my blog post table which is my mental model :grin: